packages/flutter/lib/src/services/platform_protocols.dart

```diff
@@ -0,0 +1,219 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+import 'platform_messages.dart';
+
+import 'package:flutter/foundation.dart';
+
+/// Protocols for platform communication, implemented on top of
+/// [PlatformMessages].
+class PlatformProtocols {
+  /// Returns a [Future] representing an asynchronous invocation of a named
+  /// platform function with JSON-encoded arguments, results, and errors.
+  ///
+  /// If specified, the optional [args] map must contain valid JSON values
+  /// according to [PlatformMessages.sendJSON].
+  ///
+  /// The platform function must be implemented on the native side by a handler
+  /// registered for the specified [channel]. The handler should expect
+  /// invocations using envelopes of the form
+  ///
+  ///     { "name": <function name>,
+  ///       "args": <arguments json>
+  ///     }
+  ///
+  /// If [name] is not specified, [channel] is used in its place.
+  ///
+  /// The native implementation must wrap the result of a successful invocation
+  /// in an envelope of the form
+  ///
+  ///     { "status": "ok",
+  ///       "data": <result>
+  ///     }
+  ///
+  /// Similarly, an error must be wrapped in an envelope of the form
+  ///
+  ///     { "status": "error",
+  ///       "message": <human-readable error message>,
+  ///       "data": <error details>
+  ///     }
+  ///
+  /// In both cases, `data` is optional and treated as `null` when missing.
+  /// The other envelope fields are mandatory. Additional fields, if any, are
+  /// silently discarded.
+  ///
+  /// On successful invocation, the specified [decoder] is used to turn the
+  /// result data into a [T] instance (or `null`), which will then complete
+  /// the returned future. A `null` decoder is treated as a constantly `null`
+  /// function.
+  ///
+  /// On errors, the returned future will complete with a [PlatformException]
+  /// containing the error message and error details data.
+  ///
+  /// Malformed JSON or envelopes, or data that the [decoder] cannot handle,
+  /// will make the returned Future complete with a [FormatException].
+  static Future<T> invokeJSONFunction<T>({
```

Comment:
The JSON part seems like an implementation detail, also. We'll probably have a similar wrapper on the other side, no?